### PR TITLE
Update the kit to support CRDs

### PR DESCRIPTION
**Pull request by travisn**
In K8s 1.7 CRDs were added as an enhancement and replacement for TPRs. The operator kit now supports CRDs for 1.7+ and TPRs for pre-1.7.

### CODE REVIEW

**Comment by travisn** [comment]
@jamiehannaford here are the updates to the operator kit we discussed

File: sample-operator/README.md [hunk 1]
```diff
@@ -38,11 +38,11 @@ $ kubectl get pod -l app=sample-operator
 NAME                              READY     STATUS    RESTARTS   AGE
 sample-operator-821691060-m5vqp   1/1       Running   0          3m
```
Comment (jamiehannaford) [inline sample-operator/README.md]: maybe we can nest this in the Start the Operator section so folks can verify it's running before creating the CRD

File: sample-operator/README.md [hunk 2]
```diff
@@ -38,11 +38,11 @@ $ kubectl get pod -l app=sample-operator
...
 
-$ kubectl get thirdpartyresource sample.mycompany.io
-NAME                  DESCRIPTION                             VERSION(S)
-sample.mycompany.io   A very simple example custom resource   v1alpha1
+$ kubectl get crd samples.mycompany.io
```
Comment (jamiehannaford) [inline sample-operator/README.md]: same here

File: sample-operator/sample-operator.yaml [hunk 3]
```diff
@@ -25,7 +25,6 @@ rules:
 - apiGroups:
```
Comment (jamiehannaford) [inline sample-operator/sample-operator.yaml]: I'm wondering if this ClusterRole is a bit too permissive. for example, should it be able to read every secret in the cluster? delete nodes? create namespaces? etc. the etcd-operator specifies these permissions:

https://github.com/coreos/etcd-operator/blob/master/doc/user/rbac.md#create-clusterrole

File: sample-operator/README.md [hunk 4]
```diff
@@ -64,5 +64,5 @@ $ kubectl logs -l app=sample-operator
 ```bash
 kubectl delete -f sample-resource.yaml
 kubectl delete -f sample-operator.yaml
-kubectl delete thirdpartyresource sample.mycompany.io
+kubectl delete crd samples.mycompany.io
 ```
```
Comment (jamiehannaford) [inline sample-operator/README.md]: also, it'd be nice if we can make use of multi-stage docker builds to remove the glide/golang host dependency. but that's a later task :)

Comment (bassam) [reply sample-operator/README.md]: @jamiehannaford also multi-stage docker is not widely available. when it is we'd likely use it on the rook project too.

File: resource.go [hunk 5]
```diff
@@ -18,144 +16,199 @@ limitations under the License.
...
+	if err != nil {
+		return fmt.Errorf("Error getting server version: %v", err)
+	}
+	kubeVersion := version.MustParseSemantic(serverVersion.GitVersion)
+
+	if kubeVersion.AtLeast(version.MustParseSemantic(serverVersionV170)) {
+		for _, resource := range resources {
+			err = createCRD(context, resource)
+			if err != nil {
+				return fmt.Errorf("failed to create resource %s. %+v", resource.Name, err)
```
[the remark this answers is not in the document]
Comment (travisn) [reply resource.go]: good question, I would also expect the resources to be treated independently. I'll update this.

File: resource.go [hunk 6]
```diff
@@ -18,144 +16,199 @@ limitations under the License.
...
-	KubeHTTPCli *http.Client
-}
-
 // CustomResource is for creating a Kubernetes TPR/CRD
 type CustomResource struct {
 	// Name of the custom resource
 	Name string
 
+	// Plural of the custom resource in plural
+	Plural string
```
[the remark this answers is not in the document]
Comment (travisn) [reply resource.go]: actually i don't think we can. TPRs tried to infer it and ran into issues, so it is its own field with CRDs

File: log.go [hunk 7]
```diff
@@ -18,8 +16,10 @@ limitations under the License.
...
 which also has the apache 2.0 license.
 */
+
+// Package kit for Kubernetes operators
 package operatorkit
 
 import "github.com/coreos/pkg/capnslog"
 
-var logger = capnslog.NewPackageLogger("github.com/rook/operator-kit", "op-kit")
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", "op-kit")
```
[the remark this answers is not in the document]
Comment (travisn) [reply log.go]: right, definitely want it agnostic. i'll change it

File: client.go [hunk 8]
```diff
@@ -18,33 +16,43 @@ limitations under the License.
...
 )
 
-// NewHTTPClient creates a new http client for the operator to manage the Kubernetes cluster
-func NewHTTPClient(group string) (*rest.RESTClient, error) {
+const (
+	serverVersionV170 = "v1.7.0"
+)
+
+// NewHTTPClient creates a Kubernetes client to interact with API extensions for Custom Resources
+func NewHTTPClient(group, version string, schemeBuilder runtime.SchemeBuilder) (*rest.RESTClient, *runtime.Scheme, error) {
```
[the remark this answers is not in the document]
Comment (travisn) [reply client.go]: good pointer, I'll look at simplifying it thanks

File: resource.go [hunk 9]
```diff
@@ -18,144 +16,199 @@ limitations under the License.
...
+
+	if kubeVersion.AtLeast(version.MustParseSemantic(serverVersionV170)) {
+		for _, resource := range resources {
+			err = createCRD(context, resource)
+			if err != nil {
+				return fmt.Errorf("failed to create resource %s. %+v", resource.Name, err)
+			}
+		}
+
+		for _, resource := range resources {
```
[the remark this answers is not in the document]
Comment (travisn) [reply resource.go]: i believe we actually have an async behavior here. We created all the CRDs in the previously loop, now we are waiting for them here.

File: resource.go [hunk 10]
```diff
@@ -18,144 +16,199 @@ limitations under the License.
...
+			},
+		},
 	}
 
-	for _, resource := range resources {
-		if err := waitForCustomResourceInit(context, resource); err != nil {
-			return fmt.Errorf("failed to complete init %s. %+v", resource.Name, err)
+	_, err := context.APIExtensionClientset.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
```
[the remark this answers is not in the document]
Comment (travisn) [reply resource.go]: sure i'll add an info message

File: resource.go [hunk 11]
```diff
@@ -18,144 +16,199 @@ limitations under the License.
...
+			case apiextensionsv1beta1.NamesAccepted:
+				if cond.Status == apiextensionsv1beta1.ConditionFalse {
+					fmt.Printf("Name conflict: %v\n", cond.Reason)
+				}
+			}
+		}
+		return false, err
+	})
+	if err != nil {
+		deleteErr := context.APIExtensionClientset.ApiextensionsV1beta1().CustomResourceDefinitions().Delete(crdName, nil)
```
[the remark this answers is not in the document]
Comment (travisn) [reply resource.go]: hmm, not sure why this is here. I'll take a closer look.

File: resource_test.go [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+		Timeout:               1 * time.Second,
+	}
+
+	err := createCRD(ctx, exampleResource)
+	assert.NoError(t, err)
+
+	crdName := fmt.Sprintf("%s.%s", exampleResource.Plural, exampleResource.Group)
+	crd, err := ctx.APIExtensionClientset.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crdName, metav1.GetOptions{})
+	if err != nil {
+		panic(fmt.Errorf("CustomResource.Create: %+v", err))
```
[the remark this answers is not in the document]
Comment (travisn) [reply resource_test.go]: agreed, should not be panic

File: sample-operator/README.md (hunk 1 above)
Comment (travisn) [reply sample-operator/README.md]: good idea, will move it up

File: sample-operator/sample-operator.yaml (hunk 3 above)
Comment (travisn) [reply sample-operator/sample-operator.yaml]: yeah, we could definitely scale down the permissions in the sample

**Pull request merged**